### DISCUSSION TITLE
Using SpanCategorizer without suggester

### DISCUSSION THREAD

**Discussion by thiippal** [explosion/spaCy#10657] [Help: Coding & Implementations] [answered]
Hi!

If I've understood correctly, the [SpanCategorizer](https://spacy.io/api/spancategorizer) learns to suggest Spans and to predict their label.

However, if I train a SpanCategorizer, can I disable the [Suggester](https://spacy.io/api/spancategorizer#suggesters) and use the component *only* to predict labels for spans (e.g. sentences in a Doc)?

**Comment by thomashacker** [comment, marked as answer]
Hello!
you can't really disable the `Suggester` but instead create your own custom one (e.g. a suggester that suggests all sentences in a Doc).
We currently only have the [ngram suggester](https://github.com/explosion/spaCy/blob/master/spacy/pipeline/spancat.py#L55) which you could use as a starting point for building your own function.

**Reply by thiippal** [reply]
Thanks for the answer @thomashacker!

If I create a custom suggester, should I [register](https://spacy.io/usage/training#custom-tokenizer) the function before referring  to it in the spancat configuration, as exemplified below?

```
[components.spancat]
factory = "spancat"
max_positive = null
spans_key = "sc"
threshold = 0.5
suggester = "my_custom_suggester"
```

**Reply by thomashacker** [reply]
You can initialize your custom `suggester` in the config like this:

```
[components.spancat.suggester]
@misc = "your_suggester"
...
```

You'll have to [register your custom function](https://github.com/explosion/spaCy/blob/master/spacy/pipeline/spancat.py#L55):
`@registry.misc("your_suggester")`

**Reply by thiippal** [reply]
Thanks for the help, @thomashacker!

I wrote the following custom suggester, which I load from a file using the `--code` flag during training.

```
from typing import Iterable, Optional, cast
from spacy.util import registry
from spacy.tokens import Doc
from spacy.compat import Protocol, runtime_checkable

from thinc.api import Config, Model, get_current_ops, set_dropout_rate, Ops
from thinc.types import Ragged

@runtime_checkable
class Suggester(Protocol):
    def __call__(self, docs: Iterable[Doc], *, ops: Optional[Ops] = None) -> Ragged:
        ...

@registry.misc("spacy.sent.suggester.v1")
def build_sent_suggester():

	def sent_suggester(docs: Iterable[Doc], *, ops: Optional[Ops] = None) -> Ragged:
		
		if ops is None:
			ops = get_current_ops()

		spans = []
		lengths = []

		for doc in docs:

			for sent in doc.sents:

				spans.append(ops.xp.hstack((sent.start, sent.end - 1)))

				lengths.append(len(sent))

		lengths_array = cast(Ints1d, ops.asarray(lengths, dtype="i"))

		if len(spans) > 0:

			output = Ragged(ops.xp.vstack(spans), lengths_array)

		else:

			output = Ragged(ops.xp.zeros((0, 0), dtype="i"), lengths_array)

		return output

	return sent_suggester
```

However, when I run `spacy train`, I get the following error:

```
File "custom.py", line 27, in sent_suggester
    for sent in doc.sents:
File "spacy/tokens/doc.pyx", line 873, in sents
ValueError: [E030] Sentence boundaries unset. You can add the 'sentencizer' component to the pipeline with: `nlp.add_pipe('sentencizer')`. Alternatively, add the dependency parser or sentence recognizer, or set sentence boundaries by setting `doc[i].is_sent_start`.
```

This strikes me as odd, since I've added sentence boundaries manually to my training and testing data. Have I misunderstood how the SpanCategorizer works – I thought that the incoming Doc objects are passed to the suggester, before making predictions about their labels. Any ideas what I'm doing wrong here?

**Reply by thomashacker** [reply]
I think, the problem here is that the sentences from the reference docs aren't set on the prediction docs in the [Examples](https://spacy.io/api/example). (An `Example` object holds two docs, one for prediction and one for reference/evaluation). That's why there are no sentence boundaries set for your suggester. There are two solutions I can think of to fix that:

1. Create a custom [corpus reader](https://spacy.io/api/top-level/#corpus) (a function that creates the `Example` objects) that applies the `.sent` to the prediction & reference docs
2. Or train a [senter](https://spacy.io/api/sentencerecognizer) component and add it to the pipeline + also add it to `annotating_components` in the `config.cfg` so that the sentence boundaries are set during training

**Reply by thiippal** [reply]
Does the first option involve setting the parameter `gold_preproc = true` under the corpora configuration?

If I set this parameter to true, I get a different error:

```
ValueError: [E143] Labels for component 'spancat' not initialized. This can be fixed by calling add_label, or by providing a representative batch of examples to the component's `initialize` method.
```

This seems strange to me, because I just checked that my training data contains the spans and their labels.

**Reply by adrianeboyd** [reply]
No, `gold_preproc = true` is just related to gold tokenization, not other annotation layers. For this you'd need a custom corpus reader that copies `token.is_sent_start` from the reference docs to the predicted docs. It would look a bit like this, just with `pred = nlp.make_doc(gold.text)` and `token.is_sent_start` instead of `doc.ents`:

https://github.com/explosion/projects/blob/20fd27af5122d87151e3df6277a1d6c2c3267412/tutorials/rel_component/scripts/custom_functions.py

**Reply by thiippal** [reply]
Thanks for the tip @adrianeboyd, that was easy!

My next problem: spaCy raises an IndexError when beginning to train:

```
File "/Users/thiippal/nlp/lib/python3.8/site-packages/thinc/layers/reduce_last.py", line 17, in forward
    Y = cast(OutT, Xr.dataXd[ends]) # type: ignore
IndexError: index 4507 is out of bounds for axis 0 with size 4487
```

Any ideas on how I should start debugging this?

**Reply by adrianeboyd** [reply]
It's hard to know for sure, but my first guess would be that something is wrong with the suggested spans `Ragged`. Try doing something really simple like suggesting the first word or first sentence in each doc and then expand to more of the annotation as smaller steps work. The suggester functions and `Ragged` can be a little tricky, especially around empty docs.

**Reply by thiippal** [reply]
Thanks for the help @adrianeboyd, I managed to resolve the IndexError – my `lengths_array` for the Ragged object did not have the correct shape. I managed to fix this by tinkering with @thomashacker's `spancat.py` to examine the dtypes! 🔍
